okapi-core/src/main/java/org/folio/okapi/managers/DiscoveryManager.java

```diff
@@ -166,196 +126,127 @@ void addAndDeploy(DeploymentDescriptor dd,
    *   3: No nodeId: Do not deploy at all, just record the existence (URL and instId) of the module.
    * </p>
    */
-  private void addAndDeploy0(DeploymentDescriptor dd,
-                             Handler<ExtendedAsyncResult<DeploymentDescriptor>> fut) {
+  private Future<DeploymentDescriptor> addAndDeploy0(DeploymentDescriptor dd) {
 
     String tmp = Json.encodePrettily(dd);
     logger.info("addAndDeploy: {}", tmp);
     final String modId = dd.getSrvcId();
     if (modId == null) {
-      fut.handle(new Failure<>(ErrorType.USER, messages.getMessage("10800")));
-      return;
+      return Future.failedFuture(new OkapiError(ErrorType.USER, messages.getMessage("10800")));
     }
-    moduleManager.get(modId, gres -> {
-      if (gres.failed()) {
-        if (gres.getType() == ErrorType.NOT_FOUND) {
-          fut.handle(new Failure<>(ErrorType.NOT_FOUND, messages.getMessage("10801", modId)));
-        } else {
-          fut.handle(new Failure<>(gres.getType(), gres.cause()));
-        }
-      } else {
-        addAndDeploy1(dd, gres.result(), fut);
-      }
-    });
+    return moduleManager.get(modId).compose(gres -> addAndDeploy1(dd, gres));
   }
 
-  private void addAndDeploy1(DeploymentDescriptor dd, ModuleDescriptor md,
-                             Handler<ExtendedAsyncResult<DeploymentDescriptor>> fut) {
+  private Future<DeploymentDescriptor> addAndDeploy1(DeploymentDescriptor dd, ModuleDescriptor md) {
 
     LaunchDescriptor launchDesc = dd.getDescriptor();
     final String nodeId = dd.getNodeId();
     if (nodeId == null) {
       if (launchDesc == null) { // 3: externally deployed
         if (dd.getInstId() == null) {
-          fut.handle(new Failure<>(ErrorType.USER, messages.getMessage("10802")));
+          return Future.failedFuture(new OkapiError(ErrorType.USER, messages.getMessage("10802")));
         } else {
-          add(dd, res -> { // just add it
-            if (res.failed()) {
-              fut.handle(new Failure<>(res.getType(), res.cause()));
-            } else {
-              fut.handle(new Success<>(dd));
-            }
-          });
+          return add(dd).compose(x -> Future.succeededFuture(dd));
```

Comment:
```suggestion
          return add(dd).map(dd);
```

Reply:
OK. Should be using map (which I haven't done before , I think)